docs/adr/0002-python-dependency-management-approach.md

```diff
@@ -0,0 +1,37 @@
+# Python dependency management approach
+
+## Status
+
+Proposed.
+
+## Context
+
+This ADR concerns a core element of [0001-pipeline-to-publish-python-packages](.\0001-pipeline-to-publish-python-packages.md)
+
+There are two broad approaches for managing package dependencies with Python:
```

Comment:
I once tied myself in knots once by installing a basic Python setup on my machine and then later installing Anaconda, with the result that nothing Pythonesque ever seemed to work quite right on that machine until the hard disk died and I reinstalled Windows... This led to me trying to use containers where possible to do Python work, because it lets you isolate your worlds. Given how fashionable containerisation is these days, I was guessing that this was probably a fairly common approach these days, but is that not the case?

Another point in favour of working in containers is the fact that some Python packages fire up the C compiler and build themselves some native components to get their work done. So "install package" equates more or less to "enable the stranger who wrote this package to do more or less whatever they please with my user accounts". Installing a package into a container limits the blast radius. So even if it's not common practice in the Python data science community as a whole, it is arguably something that companies who work with multiple clients should be doing to ensure isolation. Should we consider standardising on containerisation for that reason alone? (And if not, how do we ensure the safety of our customers' data?)

Reply:
I steered clear of Anaconda last year as it just installs a whole bunch of bloat on your machine and introduces specific ways of working that you have to learn on top of just getting on with using vanilla Python.

The containerised approach sounds like an option worth considering for local development and build server.  It would definitely speed up the CICD process - the creation of a Python virtual environment and installation of dependencies on the build server in my simple "sand pit" project on DevOps example currently takes about 2 minutes which I suspect would be largely redundant based on a containerised approach.

My concern is then developing a process to manage the container images themselves to keep up to date with new versions of Python or the packages.  But I guess that could be built into the scope of the pipeline.

Reply:
If we'd need to do a lot of extra research to enable this, and if it's not a thing being done broadly in the Python community then perhaps we should steer clear of containerisation. I just wanted to check whether it was a commonly used practice yet.

Reply:
Containerisation of the build is clearly an option, though typically this is done because the workload itself will be hosted in a container runtime of some sort. Where this is not the case, it can be more awkward to get the ‘outputs’ out of the build container so they can be used ‘natively’ as it were.

Whilst there is a case for isolating the package installation process for the reasons mentioned, we don’t do this for .NET and NuGet packages are arguably just as capable of running arbitrary code as part of their installation process.

We should look to bake-in the use a python package security tool from the outset.